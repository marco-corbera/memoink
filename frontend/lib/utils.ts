import { clsx, type ClassValue } from 'clsx'
import { twMerge } from 'tailwind-merge'

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}

export function getCategoryName(category: string) {
  switch (category) {
    case 'RDM':
      return 'Random Thought'
    case 'PSL':
      return 'Personal'
    case 'SCL':
      return 'School'
    case 'DRM':
      return 'Drama'
    default:
      return category
  }
}
