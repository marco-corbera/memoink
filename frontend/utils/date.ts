export function formatDate(date: Date): string {
  const inputDate = new Date(date)
  const now = new Date()
  const yesterday = new Date(now)
  yesterday.setDate(yesterday.getDate() - 1)

  if (isSameDay(inputDate, now)) {
    return 'today'
  }

  if (isSameDay(inputDate, yesterday)) {
    return 'yesterday'
  }

  return inputDate.toLocaleDateString('en-US', {
    month: 'long',
    day: 'numeric',
  })
}

function isSameDay(date1: Date, date2: Date): boolean {
  return (
    date1.getDate() === date2.getDate() &&
    date1.getMonth() === date2.getMonth() &&
    date1.getFullYear() === date2.getFullYear()
  )
}
