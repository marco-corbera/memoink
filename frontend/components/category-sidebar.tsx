"use client"

import { useEffect, useState } from "react"
import { CATEGORIES, CATEGORY_COLORS } from "@/lib/constants"
import { cn } from "@/lib/utils"
import Link from "next/link"
import { useSearchParams } from "next/navigation"
import { getCategorySummary } from "@/services/notes"
import { getCategoryName } from "@/lib/utils"

export function CategorySidebar() {
  const searchParams = useSearchParams()
  const [categoryCounts, setCategoryCounts] = useState<Record<string, number>>({})

  useEffect(() => {
    const fetchCategoryCounts = async () => {
      try {
        const summary = await getCategorySummary()
        const counts = summary.reduce((acc, { category, count }) => {
          acc[category] = count
          return acc
        }, {} as Record<string, number>)
        setCategoryCounts(counts)
      } catch (error) {
        console.error("Failed to fetch category counts", error)
      }
    }

    fetchCategoryCounts()
  }, [])

  return (
    <div className="w-64 p-6">
      <nav className="space-y-2">
        <Link
          href="/notes"
          className={cn("block px-4 py-2 rounded-lg text-memoink-button", !searchParams.get("category") && "font-medium")}
        >
          All Categories
        </Link>
        {CATEGORIES.map((category) => (
          <Link
            key={category}
            href={`/notes/category/${category}`}
            className={cn(
              "flex items-center justify-between px-4 py-2 rounded-lg text-memoink-button",
              searchParams.get("category") === category && "font-medium"
            )}
          >
            <div className="flex items-center">
              <span className={cn("w-2 h-2 rounded-full mr-2", CATEGORY_COLORS[category])} />
              {getCategoryName(category)}
            </div>
            <span className="text-sm text-memoink-button/70">{categoryCounts[category] || 0}</span>
          </Link>
        ))}
      </nav>
    </div>
  )
}

