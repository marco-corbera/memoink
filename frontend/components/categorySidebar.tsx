"use client"

import { useEffect, useState } from "react"
import { CATEGORIES, CATEGORY_COLORS } from "@/lib/constants"
import { cn } from "@/lib/utils"
import Link from "next/link"
import { usePathname } from "next/navigation"
import { getCategorySummary } from "@/services/notes"
import { getCategoryName } from "@/lib/utils"
import { Menu, X } from "lucide-react"

export function CategorySidebar() {
  const pathname = usePathname()
  const [categoryCounts, setCategoryCounts] = useState<Record<string, number>>({})
  const [isOpen, setIsOpen] = useState(false) // Estado para abrir/cerrar en mobile

  useEffect(() => {
    const fetchCategoryCounts = async () => {
      try {
        const summary = await getCategorySummary()
        const counts = summary.reduce((acc, { category, count }) => {
          acc[category] = count
          return acc
        }, {} as Record<string, number>)
        setCategoryCounts(counts)
      } catch (error) {
        console.error("Failed to fetch category counts", error)
      }
    }

    fetchCategoryCounts()
  }, [])

  return (
    <>
      {/* Botón para abrir sidebar en mobile */}
      <button
        onClick={() => setIsOpen(true)}
        className="md:hidden fixed top-4 left-4 z-50 bg-memoink-background p-2 rounded-md"
      >
        <Menu className="w-6 h-6 text-memoink" />
      </button>

      {/* Overlay en mobile */}
      {isOpen && (
        <div
          className="fixed inset-0 bg-black/50 z-40 md:hidden"
          onClick={() => setIsOpen(false)}
        />
      )}

      {/* Sidebar */}
      <div
        className={cn(
          "w-80 p-6 bg-memoink-background fixed md:static top-0 left-0 h-full md:h-auto z-50 transition-transform md:translate-x-0",
          isOpen ? "translate-x-0 shadow-lg" : "-translate-x-full md:translate-x-0"
        )}
      >
        {/* Botón para cerrar en mobile */}
        <div className="flex justify-end md:hidden">
          <button onClick={() => setIsOpen(false)} className="p-2">
            <X className="w-6 h-6 text-black" />
          </button>
        </div>

        <nav className="space-y-2">
          <Link
            href="/notes"
            className={cn("block px-4 py-2 rounded-lg text-black", pathname === "/notes" && "font-bold")}
          >
            All Categories
          </Link>
          {CATEGORIES.map((category) => {
            const categoryPath = `/notes/category/${category}`
            return (
              <Link
                key={category}
                href={categoryPath}
                className={cn(
                  "flex items-center justify-between px-4 py-2 rounded-lg text-black",
                  pathname === categoryPath && "font-bold"
                )}
                onClick={() => setIsOpen(false)} // Cierra el sidebar en mobile
              >
                <div className="flex items-center">
                  <span className={cn("w-2 h-2 rounded-full mr-2", CATEGORY_COLORS[category])} />
                  {getCategoryName(category)}
                </div>
                <span className="text-sm text-black/70">{categoryCounts[category] || 0}</span>
              </Link>
            )
          })}
        </nav>
      </div>
    </>
  )
}
