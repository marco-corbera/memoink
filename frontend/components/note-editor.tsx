"use client"

import type { Note, Category } from "@/types/note"
import { CATEGORIES, CATEGORY_COLORS } from "@/lib/constants"
import { cn } from "@/lib/utils"
import { X } from "lucide-react"
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@/components/ui/select"
import { useState, useEffect } from "react"

interface NoteEditorProps {
  note: Note
  onClose: () => void
  onUpdate: (note: Note) => void
}

export function NoteEditor({ note: initialNote, onClose, onUpdate }: NoteEditorProps) {
  const [note, setNote] = useState(initialNote)

  useEffect(() => {
    const timer = setTimeout(() => {
      onUpdate(note)
    }, 500)

    return () => clearTimeout(timer)
  }, [note, onUpdate])

  const formatLastEdited = (date: Date) => {
    return date.toLocaleString("en-US", {
      month: "long",
      day: "numeric",
      year: "numeric",
      hour: "numeric",
      minute: "2-digit",
      hour12: true,
    })
  }

  return (
    <div className="fixed inset-0 z-50 p-8 bg-black/20 backdrop-blur-sm">
      <div
        className={cn(
          "w-full max-w-5xl mx-auto h-[calc(100vh-4rem)] rounded-3xl overflow-hidden",
          CATEGORY_COLORS[note.category],
        )}
      >
        <div className="p-6 space-y-8">
          <div className="flex items-center justify-between">
            <Select
              value={note.category}
              onValueChange={(value: Category) => {
                setNote((prev) => ({
                  ...prev,
                  category: value,
                  lastEdited: new Date(),
                }))
              }}
            >
              <SelectTrigger className="w-48 border-memoink-button">
                <div className="flex items-center gap-2">
                  <span className={cn("w-2 h-2 rounded-full", CATEGORY_COLORS[note.category])} />
                  <SelectValue />
                </div>
              </SelectTrigger>
              <SelectContent>
                {CATEGORIES.map((category) => (
                  <SelectItem key={category} value={category}>
                    <div className="flex items-center gap-2">
                      <span className={cn("w-2 h-2 rounded-full", CATEGORY_COLORS[category])} />
                      {category}
                    </div>
                  </SelectItem>
                ))}
              </SelectContent>
            </Select>

            <div className="flex items-center gap-4">
              <span className="text-sm text-memoink-text/70">Last Edited: {formatLastEdited(note.lastEdited)}</span>
              <button
                onClick={onClose}
                className="text-memoink-button hover:bg-memoink-button/5 p-2 rounded-full transition-colors"
              >
                <X className="w-5 h-5" />
              </button>
            </div>
          </div>

          <div className="space-y-4">
            <input
              type="text"
              value={note.title}
              onChange={(e) => {
                setNote((prev) => ({
                  ...prev,
                  title: e.target.value,
                  lastEdited: new Date(),
                }))
              }}
              placeholder="Note Title"
              className="w-full text-4xl font-serif text-[#1A1A1A] bg-transparent border-none outline-none placeholder:text-[#1A1A1A]/40"
            />
            <textarea
              value={note.content}
              onChange={(e) => {
                setNote((prev) => ({
                  ...prev,
                  content: e.target.value,
                  lastEdited: new Date(),
                }))
              }}
              placeholder="Pour your heart out..."
              className="w-full h-[calc(100vh-280px)] bg-transparent border-none outline-none resize-none text-[#1A1A1A]/80 placeholder:text-[#1A1A1A]/40"
            />
          </div>
        </div>
      </div>
    </div>
  )
}

