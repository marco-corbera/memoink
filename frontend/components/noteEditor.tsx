'use client'

import type { Note, Category } from '@/types/note'
import {
  CATEGORIES,
  CATEGORY_COLORS,
  CATEGORY_COLORS_BORDER,
} from '@/lib/constants'
import { cn, getCategoryName } from '@/lib/utils'
import { X } from 'lucide-react'
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from '@/components/ui/select'
import { useState, useEffect, useCallback } from 'react'
import { debounce } from 'lodash'
import { updateNote } from '@/services/notes'

interface NoteEditorProps {
  note: Note
  onClose: () => void
  onUpdate: (note: Note) => void
}

export function NoteEditor({
  note: initialNote,
  onClose,
  onUpdate,
}: NoteEditorProps) {
  const [note, setNote] = useState(initialNote)
  const [hasChanges, setHasChanges] = useState(false)

  const debouncedUpdate = useCallback(
    debounce(async (updatedNote: Note) => {
      try {
        const savedNote = await updateNote(updatedNote.id, {
          title: updatedNote.title,
          content: updatedNote.content,
          category: updatedNote.category,
        })
        onUpdate(savedNote)
        setHasChanges(false)
      } catch (error) {
        console.error('Error updating note:', error)
      }
    }, 500),
    [onUpdate]
  )

  useEffect(() => {
    if (hasChanges) {
      debouncedUpdate(note)
    }
    return () => debouncedUpdate.cancel()
  }, [note, hasChanges, debouncedUpdate])

  const handleChange = (changes: Partial<Note>) => {
    setNote((prev) => ({ ...prev, ...changes, lastEdited: new Date() }))
    setHasChanges(true)
  }

  const formatLastEdited = (date: Date) => {
    if (typeof date === 'string') {
      date = new Date(date)
    }
    return date.toLocaleString('en-US', {
      month: 'long',
      day: 'numeric',
      year: 'numeric',
      hour: 'numeric',
      minute: '2-digit',
      hour12: true,
    })
  }

  return (
    <div className="w-full h-full bg-memoink-background">
      <div className="flex w-full justify-between">
        <Select
          value={note.category}
          onValueChange={(value: Category) => handleChange({ category: value })}
        >
          <SelectTrigger className="border border-memoink-text bg-memoink-background text-sm rounded-[6px] py-2 w-[225px]">
            <div className="flex items-center gap-2">
              <SelectValue />
            </div>
          </SelectTrigger>
          <SelectContent>
            {CATEGORIES.map((category) => (
              <SelectItem key={category} value={category}>
                <div className="flex items-center gap-2">
                  <span
                    className={cn(
                      'w-3 h-3 rounded-full',
                      CATEGORY_COLORS[category]
                    )}
                  />
                  {getCategoryName(category)}
                </div>
              </SelectItem>
            ))}
          </SelectContent>
        </Select>
        <button
          onClick={onClose}
          className="text-memoink-button hover:bg-memoink-button/10 p-2 rounded-full transition-colors"
        >
          <X className="w-6 h-6" />
        </button>
      </div>

      <div
        className={cn(
          'w-full h-auto bg-white rounded-[11px] shadow-md bg-opacity-50 p-8',
          `border-[3px] ${CATEGORY_COLORS_BORDER[note.category]}`,
          CATEGORY_COLORS[note.category]
        )}
      >
        <div className="flex flex-col h-full">
          <div className="mt-4 text-xs text-black self-end">
            Last Edited: {formatLastEdited(note.lastEdited)}
          </div>
          <input
            type="text"
            value={note.title}
            onChange={(e) => handleChange({ title: e.target.value })}
            placeholder="Note Title"
            className="w-full text-4xl font-serif text-black bg-transparent border-none outline-none placeholder:text-memoink-text/40"
          />
          <textarea
            value={note.content}
            onChange={(e) => handleChange({ content: e.target.value })}
            placeholder="Pour your heart out..."
            className="flex-1 bg-transparent text-black/80 placeholder:text-black/40 border-none outline-none resize-none leading-relaxed"
          />
        </div>
      </div>
    </div>
  )
}
