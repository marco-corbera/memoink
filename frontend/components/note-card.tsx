import type { Note } from "@/types/note"
import { formatDate } from "@/utils/date"
import { CATEGORY_COLORS } from "@/lib/constants"
import { cn } from "@/lib/utils"

interface NoteCardProps {
  note: Note
  onClick?: () => void
}

export function NoteCard({ note, onClick }: NoteCardProps) {
  return (
    <div onClick={onClick} className={cn("p-6 rounded-2xl cursor-pointer", CATEGORY_COLORS[note.category])}>
      <div className="space-y-3">
        <div className="flex items-center gap-2 text-sm">
          <span className="font-medium">{formatDate(note.lastEdited)}</span>
          <span className="text-memoink-text/60">·</span>
          <span>{note.category}</span>
        </div>
        <h3 className="text-2xl font-serif text-[#1A1A1A] leading-tight">{note.title}</h3>
        <div className="text-[#1A1A1A]/80">
          {note.content.includes("•") ? (
            <ul className="list-none space-y-1">
              {note.content.split("\n").map(
                (line, index) =>
                  line.trim() && (
                    <li key={index} className="line-clamp-1">
                      {line.trim()}
                    </li>
                  ),
              )}
            </ul>
          ) : (
            <p className="line-clamp-3">{note.content}</p>
          )}
        </div>
      </div>
    </div>
  )
}

