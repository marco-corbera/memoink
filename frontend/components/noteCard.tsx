import type { Note } from '@/types/note'
import { formatDate } from '@/utils/date'
import { CATEGORY_COLORS } from '@/lib/constants'
import { cn } from '@/lib/utils'
import { getCategoryName } from '@/lib/utils'

interface NoteCardProps {
  note: Note
  onClick?: () => void
}

export function NoteCard({ note, onClick }: NoteCardProps) {
  return (
    <div
      onClick={onClick}
      className={cn(
        'p-6 rounded-2xl cursor-pointer h-96 w-full overflow-hidden',
        CATEGORY_COLORS[note.category]
      )}
    >
      <div className="space-y-3">
        <div className="flex items-center gap-2 text-sm">
          <span className="font-medium">{formatDate(note.lastEdited)}</span>
          <span className="text-memoink-text/60">·</span>
          <span>{getCategoryName(note.category)}</span>
        </div>
        <h3 className="text-2xl font-serif text-[#1A1A1A] leading-tight">
          {note.title}
        </h3>
        <div className="text-[#1A1A1A]/80 overflow-hidden">
          {note.content.includes('•') ? (
            <ul className="list-none space-y-1">
              {note.content.split('\n').map(
                (line, index) =>
                  line.trim() && (
                    <li key={index} className="line-clamp-1">
                      {line.trim().slice(0, 100)}
                      {line.trim().length > 100 && '...'}
                    </li>
                  )
              )}
            </ul>
          ) : (
            <p className="line-clamp-3 overflow-hidden">
              {note.content.slice(0, 300)}
              {note.content.length > 300 && '...'}
            </p>
          )}
        </div>
      </div>
    </div>
  )
}
