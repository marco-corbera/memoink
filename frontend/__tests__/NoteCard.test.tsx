import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { NoteCard } from '@/components/NoteCard'
import type { Note } from '@/types/note'
import { formatDate } from '@/utils/date'
import { getCategoryName } from '@/lib/utils'

const mockNote: Note = {
  id: '1',
  title: 'Sample Note',
  content: 'This is a sample note content.',
  category: 'PSL',
  lastEdited: new Date(),
}

describe('NoteCard', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<NoteCard note={mockNote} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders note title', () => {
    render(<NoteCard note={mockNote} />)
    expect(screen.queryByText(mockNote.title)).toBeInTheDocument()
  })

  it('renders formatted date', () => {
    render(<NoteCard note={mockNote} />)
    expect(
      screen.queryByText(formatDate(mockNote.lastEdited))
    ).toBeInTheDocument()
  })

  it('renders category name', () => {
    render(<NoteCard note={mockNote} />)
    expect(
      screen.queryByText(getCategoryName(mockNote.category))
    ).toBeInTheDocument()
  })

  it('renders note content', () => {
    render(<NoteCard note={mockNote} />)
    expect(
      screen.queryByText(mockNote.content.slice(0, 300))
    ).toBeInTheDocument()
  })
})
