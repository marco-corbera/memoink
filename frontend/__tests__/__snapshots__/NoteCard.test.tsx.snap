// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-6 rounded-2xl cursor-pointer h-96 w-full overflow-hidden bg-category-personal"
  >
    <div
      class="space-y-3"
    >
      <div
        class="flex items-center gap-2 text-sm"
      >
        <span
          class="font-medium"
        >
          today
        </span>
        <span
          class="text-memoink-text/60"
        >
          Â·
        </span>
        <span>
          Personal
        </span>
      </div>
      <h3
        class="text-2xl font-serif text-[#1A1A1A] leading-tight"
      >
        Sample Note
      </h3>
      <div
        class="text-[#1A1A1A]/80 overflow-hidden"
      >
        <p
          class="line-clamp-3 overflow-hidden"
        >
          This is a sample note content.
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
